### PR TITLE
input: gpio_qdec: add power management support

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Add power management support to the gpio-qdec driver.

This is a bit complicated by the fact that the driver has two modes of operation and the interrupt, timer and idle work ineract with each other.

The suspend sequence is:
- set the suspended bit (inhibits the poll timer so that it does not resubmit the idle work)
- cancel the idle work (so that it does not schedule and re-set the interrupt or timers)
- disable interrupts (if used)
- stop the sampling timer
- disconnect the pins

The resume sequence is more or less the opposite.

### CODE REVIEW

**Review by bjarki-andreasen — DISMISSED** [review]
Looks good

File: drivers/input/input_gpio_qdec.c [hunk 1]
```diff
@@ -157,6 +162,12 @@ static void gpio_qdec_sample_timer_timeout(struct k_timer *timer)
 	unsigned int key;
 	uint8_t step;
 
+#ifdef CONFIG_PM_DEVICE
+	if (atomic_get(&data->suspended) == 1) {
+		return;
+	}
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:168]: is local state necessary (data->suspended)?

Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c:168]: My reasoning is that without it, the timer could fire and interrupt mid-execution of M_DEVICE_ACTION_SUSPEND and reschedule the idle-work, which would partially re-init the device, and that with this check and `k_work_cancel_delayable_sync` after that the driver should be dead for sure. Does that make sense? The synchronization here is a bit tricky

**Comment by fabiobaltieri** [comment]
fixed a bug where resume would not reset the accumulator, could cause the driver to start in a half step situation and not reset on resume, which is the intended behavior)

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@gmarull ping :-)

**Review by gmarull — APPROVED** [review]

**Pull request merged**